Zend/zend_partial.c

```diff
@@ -0,0 +1,589 @@
+        }
+    }
+
+    if (partial->named) {
+        zend_array_release(named_args);
+    }
+
+    if (partial->func.common.fn_flags & ZEND_ACC_CTOR) {
+        zval_ptr_dtor(return_value);
+
+        RETVAL_OBJ(object);
+    }
+    
+    if (!fcc.function_handler) {
+        EG(trampoline).common.function_name = NULL;
+    }
+    
+    efree(fci.params);
+}
+
+void zend_partial_apply(zval *result, uint32_t info, zend_execute_data *call) {
+    zend_function *function, *prototype = NULL;
+
+    ZVAL_OBJ(result, zend_partial_new(zend_ce_closure));
+
+    zend_partial *parent, *partial = (zend_partial*) Z_OBJ_P(result);
+
+    if ((parent = zend_partial_parent(&call->This))) {
+        function  = &parent->func;
+        prototype = &parent->prototype;
+
+        partial->argc = parent->argc + ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
+
+        memcpy(partial->argv, parent->argv, parent->argc * sizeof(zval));
+
+        uint32_t arg = 0,
+                 aend = parent->argc;
+        zval *param = ZEND_CALL_ARG(call, 1),
+             *pend = param + ZEND_CALL_NUM_ARGS(call);
+
+        while (arg < aend && param < pend) {
+            if (Z_TYPE(partial->argv[arg]) == IS_UNDEF) {
+                if (!Z_ISUNDEF_P(param)) {
+                    ZVAL_COPY_VALUE(&partial->argv[arg], param);
+                }
+                param++;
+            }
+            arg++;
+        }
+
+        if ((ZEND_CALL_INFO(call) & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+            if (parent->named) {
+                partial->named = zend_array_dup(parent->named);
+
+                zend_hash_merge(partial->named, call->extra_named_params, zval_copy_ctor, true);
+            } else {
+                partial->named = call->extra_named_params;
+                GC_ADDREF(call->extra_named_params);
+            }
+        }
+    } else {
+        function = call->func;
+
+        partial->argc = ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
+
+        memcpy(partial->argv, ZEND_CALL_ARG(call, 1), partial->argc * sizeof(zval));
+        
+        if ((ZEND_CALL_INFO(call) & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+            partial->named = call->extra_named_params;
+            GC_ADDREF(call->extra_named_params);
+        }
+    }
+
+    memcpy(&partial->func, function, (function->type == ZEND_INTERNAL_FUNCTION) ?
+              sizeof(zend_internal_function) : sizeof(zend_op_array));
+
+    partial->func.common.fn_flags &= ~ZEND_ACC_CLOSURE;
+    partial->func.common.fn_flags |= ZEND_ACC_IMMUTABLE;
+    
+    if (partial->func.common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE) {
+        partial->func.common.fn_flags |= ZEND_ACC_TRAMPOLINE_PERMANENT;
+    }
+
+    if (info & ZEND_APPLY_FACTORY) {
+        ZVAL_PTR(&partial->This, Z_OBJCE(call->This));
+
+        GC_ADD_FLAGS(Z_OBJ(call->This), IS_OBJ_DESTRUCTOR_CALLED);
+        OBJ_RELEASE(Z_OBJ(call->This));
```

Comment:
So for the `new Foo(?)` case, we first create an object, then destroy it and then if the partial is invoked, create another object.

This is non-great, though I'm not sure just how non-great it is. In conjunction with https://wiki.php.net/rfc/new_in_initializers it could mean that property default values get instantiated and immediately dropped again (though arguably that's an issue with that RFC).

Reply:
If we have space for a flag, we could use IS_OBJ_CONSTRUCTOR_CALLED here and use the initial object ...